### PR TITLE
CSHARP-370 Add Cluster.ConnectAsync() and Cluster.ShutdownAsync to the API

### PR DESCRIPTION
**Pull request by collinsauve**
Continuing conversations from [mailing list](https://groups.google.com/a/lists.datastax.com/forum/#!topic/csharp-driver-user/W1GUUqVIzUg), [CSHARP-369](https://datastax-oss.atlassian.net/browse/CSHARP-369), and [CSHARP-370](https://datastax-oss.atlassian.net/browse/CSHARP-370).  Note that despite [CSHARP-579](https://datastax-oss.atlassian.net/browse/CSHARP-579) I have some evidence  indicating this problem is actually worse in v3.3.2 compared to v3.2.1 - as soon as we upgraded our QA cluster started producing `TimeoutException`s on Connect where it did not previously (may be coincidence)

Anyways,  here's a fix that includes a `ConnectAsync` and `ShutdownAsync` that seems to work within an `AsyncContext.Run`.  

Two things:

* Needs some integration tests.  Should I just duplicate some existing tests somewhere?
* I've included a copy of  `AsyncLock` from https://github.com/neosmart/AsyncLock.  This is under MIT license.  Is that compatible with this project?  I have included the MIT copyright notice at the top of that file.

### CODE REVIEW

**Comment by landim** [comment]
Hi @collinsauve ,
first of all, thanks for the contribution. I will talk with legal about the copyright. Also, I will start the review of this PR as soon as possible.
About the integration tests, I did not understood your question? What do you mean by duplicate existing?

**Comment by collinsauve** [comment]
Re: Integration tests - I'm thinking there should probably be some tests for these new `ConnectAsync` and `ShutdownAsync` functions.   Are there tests already for `Connect` and `Shutdown` that I can duplicate for the async versions?

**Comment by landim** [comment]
Hi @collinsauve you can take a look at src/Cassandra.IntegrationTests/Core/ConnectionTests.cs. Although, basically all tests use Connect method on TestSetup.

**Comment by collinsauve** [comment]
HI @landim I explorered the tests more and found `ClusterTests`.  I've parameterized those tests to also attempt connect/shutdown using the new `ConnectAsync` and `ShutdownAsync` functions.  Let me know if this is wrong/I should add more tests.

P.S. I will be on vacation next week so will not be very responsive, if at all. :)

**Comment by jorgebay** [comment]
Enjoy your vacations! When you come back, we can discuss if we can avoid using `AsyncLock` in favor of `SemaphoreSlim.WaitAsync()`.
I think its possible to use just `await semaphore.WaitAsync()` and then `semaphore.Release()` wrapped in a `try...finally` block, wdyt?

**Comment by collinsauve** [comment]
I believe the benefit of this `AsyncLock` implementation is that it is re-entrant safe.  So long as we do not need that, `SemaphoreSlim` should work just fine.

File: src/Cassandra/Cluster.cs [hunk 1]
```diff
@@ -39,7 +39,7 @@ public class Cluster : ICluster
         private readonly ControlConnection _controlConnection;
         private volatile bool _initialized;
         private volatile Exception _initException;
-        private readonly object _initLock = new object();
+        private readonly AsyncLock _initLock = new AsyncLock();
```
Comment (jorgebay) [inline src/Cassandra/Cluster.cs]: Change it for a `SemaphoreSlim` instance.

**Comment by collinsauve** [comment]
I've changed it to a `SemaphoreSlim`, and squashed my commits.

**Comment by jorgebay** [comment]
I saw you made the change @collinsauve, I'll look at it in the next few days, sorry I can not look at it earlier... :)

**Comment by jorgebay** [comment]
Somehow related: I've added [CSHARP-598](https://datastax-oss.atlassian.net/browse/CSHARP-598) to simplify `PreserveStackTrace()` implementation.

File: src/Cassandra/Cluster.cs [hunk 2]
```diff
@@ -39,7 +39,8 @@ public class Cluster : ICluster
         private readonly ControlConnection _controlConnection;
         private volatile bool _initialized;
         private volatile Exception _initException;
-        private readonly object _initLock = new object();
+        private readonly SemaphoreSlim _initLock = new SemaphoreSlim(0, 1);
```
Comment (jorgebay) [inline src/Cassandra/Cluster.cs]: It should be `new SemaphoreSlim(1, 1)` as `initialCount` refers to initially available (its confusing setting!)

Comment (collinsauve) [inline src/Cassandra/Cluster.cs]: Thanks @jorgebay, fixed.

**Comment by collinsauve** [comment]
Hi @jorgebay.  Just circling back to this in case it got lost/forgotten about.   I believe I've made all the changes you requested.  Is there anything else I can do to help move this along?

**Comment by collinsauve** [comment]
Btw it appears as though your appveyor is broken, multiple PRs (including this one) are failing with the same errors.  See the last 3 builds in [appveyor history](https://ci.appveyor.com/project/DataStax/csharp-driver/history).

**Comment by jorgebay** [comment]
Sorry for the long delay, thanks for bringing this up, I think I will be able to give a final look at this patch this week.

**Comment by jorgebay** [comment]
I've done final review, focusing on finding possible deadlocks and lgtm.

Regarding the broken builds, I think its due to the several changes in the build tools we made in the past months, I've created a merge commit on a separate branch in order to trigger CI builds (appveyor and private Jenkins): https://ci.appveyor.com/project/DataStax/csharp-driver/build/connect-async-merge-1336

I've created 2 tickets to follow up on this API change: CSHARP-607 (DSE Driver) and CSHARP-608 (for `ICluster` interface, which is a breaking change).

Regarding tests, we should add some more tests for `ConnectAsync()` with paused nodes and stuff like that, cc @landim 

I'll wait for the CI builds to finish and merge it. Thanks again @collinsauve for a great contribution and for your patience!

**Pull request merged**